[프레디]미션 6. 기물 이동 구현

Dae-Hwa opened the pull request:
안녕하세요 프레디입니다!

미션 6 완료했습니다.

생각보다 속도가 많이 느려서 늦게 제출하게 됐습니다. 😥

수정과 추가된 사항이 많았는데, 너무 복잡해진 것은 아닌지 걱정이 되긴 하네요..ㅎㅎ

냉정한 리뷰 부탁드립니다. !! 

**Main**

-   `move` 명령어가 추가됐습니다.
    -   명령어는 단순 string이 아닌 `Command` 인터페이스를 추가하여 처리했습니다.
-   예외 발생 시 프로그램을 종료하지 않고, 메세지 출력 후 게임을 이어서 진행할 수 있도록 했습니다.

**Command**

-   `Main`의 명령어를 처리하기 위해 추가된 인터페이스입니다.
-   `Start`, `Move`로 나누었고, `end`의 경우 추가적인 처리가 필요하지 않다고 판단하여 클래스로 분리하지 않았습니다.

**Start**

-   게임을 시작하기 위한 `Command`의 구현입니다.

**Move**

-   기물을 이동시키는 `Command`의 구현입니다.

**Board**

-   화면을 보여주기 위한 `BoardView`와 게임을 진행하는 `ChessGame`으로 역할을 나눴습니다.

**BoardView**

-   화면을 보여주기 위한 클래스입니다. `Board`에 있던 `getRepresentation()`을 옮겨 담았습니다.

**ChessGame**

-   게임을 진행하기 위한 클래스입니다. `initialize`와 `move`를 담당합니다.
-   `isInitialized` 플래그를 추가하여 게임이 진행 상태에 따라 예외 처리를 했습니다.

**Piece**

-   기물이 이동 가능한지 판단하는 `isCanMove`를 추가했습니다. 폰을 제외한 기물들은 간단한 수식으로 처리할 수 있어 방향을 나타내는 enum을 추가하지 않고 진행했습니다. 폰의 경우도 로직이 복잡해지는 건 enum을 쓰나 현재 방식대로 하나 비슷하다고 판단되어 계속해서 진행했습니다.

-   처음에는 `Kind` 내부에 abstract 메소드로 이동 가능 여부 판단 로직을 넣었습니다. 하지만 폰의 경우 로직이 복잡해져 해당 로직을 `Piece`로 옮기고, `Piece`의 자식 클래스를 새로 생성했습니다.

-   자식 클래스에 생성자 대신 정적 팩토리 메소드를 사용하도록 했습니다. 자식 클래스의 팩토리 메소드의 리턴 타입을 추상화된 부모(혹은 인터페이스) 타입으로 해야 하는지, 해당 클래스의 타입 그대로를 리턴해야 하는지 고민하다가, 생성자 대신 사용하는 것이니 생성자처럼 해당 클래스의 타입을 리턴하도록 작성했습니다.

-   아래는 각 기물 별 설명입니다.

    **Pawn**

    1.  입력한 위치가 이동 가능한 범위 내에 있는지 확인합니다.
    2.  이동 경로가 대각선인지 확인하고 현재 대각선으로 움직일 수 있는지 확인합니다.
    3.  흑의 경우 아래로, 백의 경우 위로 움직여야 합니다.

    **Knight**

    -   x축 이동 거리와 y축 이동거리의 합은 반드시 3이어야 하고, 차의 절대값은 반드시 1이 돼야 합니다. 

    **Rook**

    -   주어진 좌표의 기울기가 반드시 0이어야 합니다

        >   y축으로 평행하게 움직이는 경우도 0이라고 판단합니다.

    **Bishop**

    -   기울기의 절대값이 반드시 1이어야 합니다.

    **Queen**

    -   기울기의 절대값이 0이거나 1입니다.

    **King**

    -   x축과 y축 이동 거리의 절대값이 각각 1을 넘으면 안됩니다.

    **Empty**

    -   항상 이동이 불가능합니다.

**PositionCalculator**

-   `Piece`의 자식 클래스들에서 재정의 한 `isCanMove`에서 반복적으로 발생하는 코드를 처리하기 위한 클래스입니다. 

-   출발점과 도착점의 위치로 거리와 기울기를 계산합니다.

**PieceFactory**

-   Piece 수정 사항에 맞춰 팩토리 메소드를 수정했습니다.


감사합니다!

Dae-Hwa committed:
Add move in Board

King 움직이기 구현

Dae-Hwa committed:
Create Command

start와 move를 처리하는 Command 인터페이스 및 Start, Move 클래스 추가

Dae-Hwa committed:
Add verifyMovePosition in Kind and move check logic in Board.move

Kind에 abstract로 verifyMovePosition 만들어 체크 로직 구현. 무브에 있던 킹 움직이는 로직 옮김.

Dae-Hwa committed:
Add verifyMovePosition of Knight in Kind

Dae-Hwa committed:
Add verifyMovePosition of BISHOP in Kind

Dae-Hwa committed:
Refactor BoardTest for remove duplicated move checking methods

Dae-Hwa committed:
Add verifyMovePosition of Rook in Kind

Dae-Hwa committed:
Add verifyMovePosition of EMPTY in Kind

Dae-Hwa committed:
Add verifyMovePosition of Queen in Kind

Dae-Hwa committed:
Refactor Kind with adding gradientOf

룩, 비숍, 퀸은 기울기를 이용하여 판단하므로 기울기 계산 메소드 추출하여 사용

Dae-Hwa committed:
Reorder test method moveKing in BoardTest

Dae-Hwa committed:
Move color checking logic in Board to verifyMovePosition in Kind

중복 코드가 발생하지만 verifyMovePosition 내부에서 판단하는게 더 적합하다고 판단.

Dae-Hwa committed:
Change the name and return type of verifyMovePosition in Kind

boolean 타입으로 변경하고, isCanMove로 이름 변경

Dae-Hwa committed:
Add isCanMove of Pawn in Kind

Dae-Hwa committed:
Rename BoardTestBase to ChessTestBase and move to package chess

Dae-Hwa committed:
Create ChessGame and move initialize and move from Board

Dae-Hwa committed:
Remove unnecessary import

Dae-Hwa committed:
Add BoardView and move getRepresentation to BoardView

Dae-Hwa committed:
Refactor if statements of Kind

Dae-Hwa committed:
Remove unnecessary import

Dae-Hwa committed:
Remove unnecessary return type

Dae-Hwa committed:
Remove createBlank in Piece

Dae-Hwa committed:
Extract subclasses from Piece and make isCanMove to abstract

Kind로 나눠놨던 종류를 클래스로 분류, Piece는 abstract로 변경하여 isCanMove를 하위 클래스에서 구현함

Dae-Hwa committed:
Refactor Pawn by extracting methods from isCanMove

Dae-Hwa committed:
Add initBoard for test in ChessTestBase

Dae-Hwa committed:
Create PositionCalculator

Dae-Hwa committed:
Apply PositionCalculator

Dae-Hwa committed:
Fix package of PositionTest and test of Position's subclasses

Dae-Hwa committed:
Remove unnecessary methods in Piece with use PositionCalculator

Dae-Hwa committed:
Move tests of ChessGame to subclasses of Piece

ChessGameTest에 있던 isCanMove와 관련된 테스트를 각 기물 클래스의
테스트로 이동

Dae-Hwa committed:
Improve console messages of Main

Dae-Hwa committed:
Add isInitialized in ChessGame

Dae-Hwa committed:
Improve console message in Main

Dae-Hwa committed:
Fix Move to throw err when argument size is wrong

ksundong commented on src/main/java/kr/codesquad/freddie/chess/Main.java:
```diff
@@ -9,16 +11,34 @@
+        System.out.print(sb.toString());
+
+        return br.readLine();
+    }
```
이 메서드가 과연 Main 클래스가 가지고 있어야 할 메서드일지 고민을 해볼까요?

ksundong replied:
또한 불필요하게 StringBuilder를 쓴 부분으로 보여지네요!

Dae-Hwa replied:
처음에 읽어올 수 있도록 해주는 클래스를 만들어봤는데, 출력해주는 소스가 섞여 어색한 것 같아 `ChessGamePrinter`를 새로 만들었습니다.
1a8225b

ksundong commented on src/main/java/kr/codesquad/freddie/chess/Main.java:
```diff
@@ -9,16 +11,34 @@
+                    System.out.println("잘못된 입력입니다. command : " + command);
+                } catch (IllegalStateException e) {
+                    System.out.println(e.getMessage());
                 }
```
try catch 부분도 메서드로 추출해보면 어떨까요?

Dae-Hwa replied:
78d0667
메소드로 추출해봤는데, 프린트 해주는 부분은 모두 위에서 만든 프린터 클래스를 활용하는게 좋을까요?
애매하다는 생각이 들었던 부분이 있어 질문을 드리는데, 
두 번째 catch문에서 현재는 메세지를 그대로 프린트하도록 되어있습니다.
이런 경우 메소드로 추출해도 크게 의미가 없지 않나 하는 생각이 들었고, 그렇다고 위쪽의 프린트 구문만 출력하는 것은 일관성이 없지 않나? 하는 생각이 들었기 때문입니다.

ksundong replied:
좋은 고민이네요. 저 같은 경우에는 책임을 한 쪽에 몰아주는 편이 더 나았던 것 같아요.

ksundong commented on src/main/java/kr/codesquad/freddie/chess/board/Board.java:
```diff
@@ -85,10 +73,10 @@ public Piece findPiece(String position) {
 
-    public Piece set(String position, Piece piece) {
+    public void set(String position, Piece piece) {
         Position positionConverter = Position.of(position);
```
이름은 Converter인데 딱히 Convert를 해주진 않네요. 😅

Dae-Hwa replied:
이전 단계에서 반영하다 빼먹은 것 같습니다. 😅 수정해줬습니다. 
934120b

ksundong commented on src/main/java/kr/codesquad/freddie/chess/board/Board.java:
```diff
@@ -127,6 +109,10 @@ public int getNumberOf(Color color, Kind kind) {
     }
 
+    public List<File> getFiles() {
+        return Collections.unmodifiableList(files);
```
요번 기수분들은 `unmodifiableList`를 잘 쓰시네요. 신기신기...

Dae-Hwa replied:
집단지성의 힘인것 같습니다! ㅋㅋ

ksundong commented on src/main/java/kr/codesquad/freddie/chess/board/BoardView.java:
```diff
@@ -0,0 +1,27 @@
+
+import java.util.stream.Collectors;
+
+public class BoardView {
+    private Board board;
```
딱히 변경되지 않을 값이라면 `final`을 붙여주어도 좋을 것 같습니다.

Dae-Hwa replied:
이외에도 final이 들어갈 수 있다면 직접 명시해주는게 더 좋을까요?
그 동안은 불변성이 소스적으로(표현이 이상하네요 ㅋㅋ) 막혀있으면 충분한 표현이 되지 않을까 했는데 다시 생각해보니 final을 붙이는게 훨씬 직관적일 것 같습니다

ksundong replied:
최근 들어서는 `final`을 직접적으로 명시해주는 경향이 짙은 것 같아요.
저도 `final`을 명시적으로 해주는 것을 선호합니다.

ksundong commented on src/main/java/kr/codesquad/freddie/chess/board/BoardView.java:
```diff
@@ -0,0 +1,27 @@
+        return result;
+    }
+
+    private String getRepresentationOf(File file) {
+        return file.getPieces().stream()
+                .map(Piece::getRepresentation)
+                .collect(Collectors.joining());
+    }
```
이 메서드를 보고 든 생각인데 Files도 클래스로 뺀다면, 이 메서드는 그곳에 위치할 수 있을 것 같네요~

Dae-Hwa replied:
getRepresentationOf라는 저 메소드 자체도 view를 위한 것이 아닌가 하는 생각이 들었습니다. 사실 이렇게 따지면 Piece의 getRepresentation도 마찬가지일 것 같은데요. 이런 경우에 외부에 노출될 최상위 클래스에서만 이런 분리를 해주면 될지, 아니면 각 개층별로 View를 위한 클래스를 각각 만들어주는게 더 좋은 것인지 궁금합니다.

Dae-Hwa replied:
그리고 Files로 분리하다보니 Board는 Files의 기능을 전달하는 역할밖에 못하는게 아닌가 하는 의문도 들었습니다. 그러다보니 이런식으로 분리하는게 괜찮은 것인지에 대한 의문이 함께 들었습니다.
아니면 제가 역할 정의나 분리를 잘못해서 그런건지도 궁금합니다.

Dae-Hwa replied:
bf3a48b

ksundong replied:
`Board`는 `Board`의 책임이 있고, `Files`는 `Files`의 책임이 있겠죠?
`Board`가 해당 역할만 수행한다면 의미가 없을지 모르겠지만, 그건 아닐텐데요!

ksundong commented on src/main/java/kr/codesquad/freddie/chess/board/Position.java:
```diff
@@ -30,4 +30,13 @@ public int getFileIndexForList() {
+                .append(getRankIndex());
+
+        return sb.toString();
+    }
```
왠만하면 `toString()`을 오버라이딩 하기보다는, `toString()`은 정말 객체의 상태를 나타낼 때만 사용하고, 별도의 메서드를 정의해주는 편이 좋을 것 같아요.
객체의 상태를 표현하려던 것이라면, 툴을 이용해 자동으로 만들어주세요!

ksundong commented on src/main/java/kr/codesquad/freddie/chess/command/Command.java:
```diff
@@ -0,0 +1,19 @@
+
+import kr.codesquad.freddie.chess.game.ChessGame;
+
+public interface Command {
```
Enum으로 만들어봐도 재밌었을 것 같아요?
좋은 추상화인지, 과도한 추상화인지는 조금 고민을 해봐야할 것 같습니다.
커맨드를 추상화 한다는건 꽤 놀라운 발상인 것 같아요. 👍

Dae-Hwa replied:
enum도 고려해봤었는데, 인터페이스로 분리했던 이유는 사용하는 매개변수가 달라질 수 있어서였습니다.
예를 들어, start의 경우 매개변수가 필요없이 그냥 초기화해주면 되겠지만 move의 경우 매개변수가 필요합니다. 그럴 경우 공통된 메소드로 사용하게되면
```java
// Start의 메소드
@Override
execute(String command) {
    command 변수를 사용하지 않음.
}

// Move의 메소드
@Override
execute(String command) {
    command 변수를 사용함
}
```
이런 식의 구조가 되기 때문에 메소드의 시그니쳐는 공통적으로 유지하되, 클래스의 필드를 다르게 지정하였습니다.
위에 있는 사용하지 않는 매개변수가 선언 된 메소드가 좋아보이지 않아서였습니다.

사실 Piece의 canMove에서도 비슷한 상황이 발생합니다.
Pawn을 판단할때는 대각선으로 갈 수 있는 경우가 색상이 다르며 빈 칸이 아닌 기물이 있어야 하는데, 이를 판단하기 위해 추가적인 매개변수가 필요했기 때문입니다. 
```java
@Override
public boolean canMove(PositionCalculator positionCalculator, Piece targetPiece) {
    targetPiece가 Pawn에서만 사용되고, 다른 기물에서는 사용되지 않음.
}
```
이는 단순히 false만 리턴해주는 CalculatalbePiece나 EmptyPiece에서는 더 심해지는데,
이럴 경우 위에서 제가 선택한 객체마다 다른 상태를 지정하여 생성해주도록 하는게 좋은 것인지, 아니면 이 정도는 크게 상관없는 경우인지 궁금합니다.
CalculatalbePiece는 조금 다른 경우인 것 같긴 하지만, https://github.com/codesquad-members-2021/java-chess/pull/166#discussion_r586301907 에서 추상화에 관한 코멘트를 남겨주셔서 공통적으로 판단이 잘 되지 않았던 부분을 좀 길게 질문드렸습니다.

ksundong replied:
🤔 어려운 문제네요. 매개변수를 안받는 구조로 만들기 위해선 어떻게 해야할지도 고민해볼만 한 것 같습니다.
Enum은 지금 생각해보니 그리 적당한 해결책은 아닌 것 같네요.

사용하지 않는 매개변수가 좋아보이지 않는다는 것은 맞습니다.

다만, 후자의 경우에는 설계의 방향을 바꾸는건 어떤지도 고민해볼만한 것 같습니다. 저라면 그냥 `canMove(Board board)` 로 처리해줬을 것 같거든요.

어쩌면 ISP를 위반하는건 아닌지 의심해보세요 ㅎㅎ

ksundong commented on src/main/java/kr/codesquad/freddie/chess/command/Command.java:
```diff
@@ -0,0 +1,19 @@
+import kr.codesquad.freddie.chess.game.ChessGame;
+
+public interface Command {
+    static Command create(String command) {
```
`create`도 좋지만, `of`도 나쁘지 않을 것 같다는 생각이 드네요.

Dae-Hwa replied:
항상 고민이 많이 되는 부분인데, 혹시 이런 네이밍할때 참고하시는 자료나 기준이 있으신가요?
저는 찾아도 잘 나오지 않아 이팩티브 자바를 참고했었는데 여기 나오는 용례랑 다르게 of를 사용하시는 분도 많으신 것 같아서요
이 책이 교과서가 아니니 당연하긴 하지만 막상 찾아보면 이 책과 무관한 자료를 찾기힘든 아이러니함이 발생합니다 😅

ksundong replied:
저는 주로 읽을 때와 관례를 많이 참고합니다.
여기서는 `create()`라기 보다는 `of()`에 가까운 느낌이 들었어요.
관례와 다르다...기 보다는 `of()`가 어디에서나 자주 사용될 수 있는 형태여서 그렇다고 생각하는게 좋을 것 같네요.

ksundong commented on src/main/java/kr/codesquad/freddie/chess/command/Move.java:
```diff
@@ -0,0 +1,21 @@
+    public Move(String[] command) {
+        if (command.length != 3) {
+            throw new IllegalArgumentException("잘못된 입력입니다." + Arrays.toString(command));
+        }
```
별도의 메서드로 빼보는건 어때요?

ksundong commented on src/main/java/kr/codesquad/freddie/chess/command/Move.java:
```diff
@@ -0,0 +1,21 @@
+    }
+
+    @Override
+    public void execute(ChessGame chessGame) {
```
호출할 때 조금 특이한 구조가 될 것 같아요. `move.execute(chessGame);`

ksundong commented on src/main/java/kr/codesquad/freddie/chess/command/Start.java:
```diff
@@ -0,0 +1,10 @@
+
+public class Start implements Command {
+    @Override
+    public void execute(ChessGame chessGame) {
```
이것저것 생각을 해봤는데 ChessGame이 Command를 들고 있고, execute할 때 마다 해당 동작을 수행하는 식으로 구현해보면 어떨까? 하는 생각이 드네요.

ksundong commented on src/main/java/kr/codesquad/freddie/chess/game/ChessGame.java:
```diff
@@ -0,0 +1,64 @@
+    }
+
+    public void initialize() {
+        if (isInitialized) {
+            throw new IllegalStateException("게임이 이미 시작되었습니다.");
+        }
```
이 부분을 메서드로 분리하면 어떨까요?

Dae-Hwa replied:
분리할 때
```java
 private void checkGameInitialized() {
    if (isInitialized) {
        throw new IllegalStateException("게임이 이미 시작되었습니다.");
    }
}
```
이런식으로 하니 또 다른 중복이 생기는 느낌인데,
```java
private void checkState(boolean isNormalState, String message) {
    if(!isNormalState) {
        throw new IllegalStateException(message);
    }
}
```
이런 식으로 분리해서 공통 유틸처럼 사용해도 괜찮은지, 아니면 체크 로직은 각각의 의미를 정확히 담아주는게 좋은 것인지 궁금합니다!

ksundong replied:
저는 미래지향적 설계는 지양합니다. 필요한 시점에 공통 로직이 존재하면 그 때 추출해요.
중복이 생기면 그 때 추출해주면 됩니다. 😊
마지막 질문은 딱히 정답이 없는 문제라 저 같은 경우에는 각각의 의미를 정확히 담아주도록 설계합니다.
코드를 읽을 때, 잘 읽힌다고 느꼈거든요.

ksundong commented on src/main/java/kr/codesquad/freddie/chess/game/ChessGame.java:
```diff
@@ -0,0 +1,64 @@
+    }
+
+    public void move(String source, String target) {
+        if (!isInitialized) {
+            throw new IllegalStateException("게임이 시작되지 않았습니다.");
+        }
```
이부분도 메서드로 빼주어도 좋을 것 같아요.

ksundong commented on src/main/java/kr/codesquad/freddie/chess/game/ChessGame.java:
```diff
@@ -0,0 +1,64 @@
+            throw new IllegalArgumentException(message);
+        }
+
+        if (!sourcePiece.isCanMove(new PositionCalculator(Position.of(source), Position.of(target)), targetPiece)) {
```
```suggestion
        if (!sourcePiece.canMove(new PositionCalculator(Position.of(source), Position.of(target)), targetPiece)) {
```

굳이 동사가 두 개 들어갈 필욘 없겠죠?

ksundong replied:
그리고 `PositionCalculator`가 객체로 생성되어야 할 지, 이 코드에서는 아직 잘 모르겠네요.

ksundong commented on src/main/java/kr/codesquad/freddie/chess/game/ChessGame.java:
```diff
@@ -0,0 +1,64 @@
+        }
+
+        board.set(target, sourcePiece);
+        board.set(source, PieceFactory.createBlank());
```
생각해보면 굳이 Blank를 매 번 새로 만들어줄 필요가 있을까 싶네요.

Dae-Hwa replied:
05296ff
상수로 선언해서 재사용했습니다.

ksundong commented on src/main/java/kr/codesquad/freddie/chess/game/ChessGame.java:
```diff
@@ -0,0 +1,64 @@
+        return board;
+    }
+
+    @Override
+    public String toString() {
+        return "ChessGame{" +
+                "board=" + board +
+                '}';
+    }
```
toString을 선언해주신 이유가 무엇일까요?

Dae-Hwa replied:
지금은 특별한 이유가 보이지는 않는 것 같습니다.
그동안 필요하다고 생각되는 클래스에만 만들어줬었는데, toString을 모든 클래스에 재정의 해주는게 권장된다는걸 봐서 그렇게 해볼까 했습니다(지금은 몇 군데만 재정의 된 상태입니다).
그렇게 하는 것 보다 꼭 필요한 곳에만 만들어주는게 더 좋은가요? 아니면 혹시 말씀해주신 것에 다른 의도가 있으신지 궁금합니다!

ksundong replied:
`toString()`을 마구만들다보면, 객체 그래프가 순환형태일 경우 `stackOverFlow`가 발생합니다.
왜 인지는 직접 경험하시면 아실텐데, 아마 곧 경험하실 것 같네요.

ksundong replied:
`toString()`을 재정의할 때, 유의미한 정보를 전달하는 것이 좋은데, 기계적으로 만드는 경우를 경계하라는 의미입니다~

Dae-Hwa replied:
아... 그런 경우가 생길 수 있겠네요.
감사합니다! 바로 이해됐습니다.

ksundong commented on src/main/java/kr/codesquad/freddie/chess/piece/Bishop.java:
```diff
@@ -0,0 +1,18 @@
+import kr.codesquad.freddie.chess.utils.PositionCalculator;
+
+public class Bishop extends Piece {
+    private Bishop(Color color, Kind kind) {
```
제가 봤을 때에는 이 생성자가 Piece쪽에 선언되었다면 어떨까? 하는 생각이 드네요.

Dae-Hwa replied:
어떤식으로 해야할지 감이 잘 오지않는데 혹시 한번만 더 설명해주실 수 있으신가요? 
제가 생각할때는 `Piece`에 생성자가 이미 있는 것 같아서요

ksundong replied:
아 이미 존재하고 있군요! 저는 개별 생성자가 없는 방향을 생각하고 있었습니다!

ksundong commented on src/main/java/kr/codesquad/freddie/chess/piece/Bishop.java:
```diff
@@ -0,0 +1,18 @@
+    }
+
+    @Override
+    public boolean isCanMove(PositionCalculator positionCalculator, Piece targetPiece) {
```
다른 곳에서 요청드린바와 마찬가지로, `canMove` 정도로도 충분할 것 같아요.

Dae-Hwa replied:
수정했습니다! 6a6dee3

ksundong commented on src/main/java/kr/codesquad/freddie/chess/piece/CalculablePiece.java:
```diff
@@ -14,6 +16,11 @@ public static CalculablePiece create(Piece piece, char fileIndex) {
+    @Override
+    public boolean isCanMove(PositionCalculator positionCalculator, Piece targetPiece) {
+        return false;
+    }
```
항상 `false`라면 어쩌면 추상화가 조금 다른 방향으로 되어야 하지 않을까? 라는 의심을 해볼만 할 것 같아요.

ksundong commented on src/main/java/kr/codesquad/freddie/chess/piece/Pawn.java:
```diff
@@ -0,0 +1,49 @@
+
+    private boolean isAvailableMoveRange(int fileDistance, int rankDistance) {
+        return Math.abs(rankDistance) <= 2 && Math.abs(fileDistance) <= 1;
+    }
```
체스 규칙상 처음 말고는 1칸만 전진할 수 있지 않나요?

Dae-Hwa replied:
놓치고 있었던 부분이네요 😅 수정했습니다. 감사합니다!
71e1dd7

ksundong commented on src/main/java/kr/codesquad/freddie/chess/utils/PositionCalculator.java:
```diff
@@ -0,0 +1,50 @@
+import kr.codesquad.freddie.chess.board.Position;
+import kr.codesquad.freddie.chess.piece.Piece;
+
+public class PositionCalculator {
```
네이밍 어렵네요...
책임을 분리해주신건 좋은 것 같아요.

ksundong replied:
👍 💯

ksundong commented on src/test/java/kr/codesquad/freddie/chess/ChessTestBase.java:
```diff
@@ -0,0 +1,24 @@
+            board.add(PieceFactory.createBlank());
+        }
+    }
+}
```
오, 공통된 조건으로 테스트 환경을 만들어 줄 필요가 있었나보네요.

ksundong commented on src/test/java/kr/codesquad/freddie/chess/board/BoardTest.java:
```diff
@@ -11,21 +12,27 @@
+        board.initializePieceBy(Color.BLACK);
+
         checkInitializeRoyal(Color.BLACK);
-        checkInitializeRoyal(Color.WHITE);
         checkInitializePawn(Color.BLACK);
```
assertAll()로 묶어줘도 좋을 것 같네요~

ksundong commented on src/test/java/kr/codesquad/freddie/chess/board/BoardTest.java:
```diff
@@ -58,31 +65,17 @@ private void checkInitializePawn(Color color) {
+        board.initializeEmptyPiece();
+
+        StringBuilder sb = new StringBuilder();
+        for (int i = 3; i <= 6; i++) {
```
3, 6은 어떤 의미인지 알기 어렵다고 생각됩니다.
그리고 되도록이면 저는 부등호만 사용하고, 등호는 가급적이면 빼주는 편이에요!

ksundong commented on src/test/java/kr/codesquad/freddie/chess/board/BoardTest.java:
```diff
@@ -58,31 +65,17 @@ private void checkInitializePawn(Color color) {
+        StringBuilder sb = new StringBuilder();
+        for (int i = 3; i <= 6; i++) {
+            sb.delete(0, sb.length());
+            for (char j = 'a'; j <= 'h'; j++) {
```
와 ㅋㅋ char를 for loop에 쓰시다니 역시 for문의 마법사시네요.
의도도 대충 알겠지만, 가급적이면 for문에는 숫자 타입을 쓰시는걸... 권해드리고 싶네요.

ksundong replied:
그리고 이중 for문 테스트 코드라서 크게 지적하고 싶진 않고, 메서드 추출을 시도해보시는게 좋을 것 같아요~

ksundong commented on src/test/java/kr/codesquad/freddie/chess/board/BoardTest.java:
```diff
@@ -58,31 +65,17 @@ private void checkInitializePawn(Color color) {
+        for (int i = 3; i <= 6; i++) {
+            sb.delete(0, sb.length());
+            for (char j = 'a'; j <= 'h'; j++) {
+                assertThat(board.findPiece(sb.append(j).append(i).toString()))
```
저는 꼭 i, j 가 아니라 다른 변수를 시도해봐도 좋다고 생각합니다!

ksundong replied:
그리고 해당 assertion은 메서드로 추출해보는 것도 시도해보세요~

ksundong commented on src/test/java/kr/codesquad/freddie/chess/board/BoardTest.java:
```diff
@@ -92,7 +85,7 @@ void add() {
 
             for (int j = 0; j < i; j++) {
-                board.add(new Piece(Color.WHITE, Kind.PAWN));
+                board.add(whitePieceFactory.createPawn());
```
Factory Class 적용하니까 좋네요.

ksundong commented on src/test/java/kr/codesquad/freddie/chess/board/BoardTest.java:
```diff
@@ -151,7 +144,7 @@ void findPieceMoreThanOne() {
     void getPiecesBy() {
-        board.initialize();
+        initBoard(board);
         assertThat(board.getPiecesBy(Color.BLACK)).isEqualTo(Arrays.asList(
```
다음 assertion은 `contains`를 활용해보는 것도 좋아보여요~

ksundong commented on src/test/java/kr/codesquad/freddie/chess/board/BoardTest.java:
```diff
@@ -151,7 +144,7 @@ void findPieceMoreThanOne() {
     @Test
     void getPiecesBy() {
-        board.initialize();
+        initBoard(board);
```
`initBoard()`가 어떤 역할을 하는 걸까요?
`initialize()`로 충분하지 않았었나요?

ksundong commented on src/main/java/kr/codesquad/freddie/chess/board/Board.java:
```diff
@@ -20,33 +20,21 @@ public Board() {
         }
     }
 
-    public Board initialize() {
-        initializePieceBy(Color.BLACK);
-        initializeEmptyPiece();
-        initializePieceBy(Color.WHITE);
-        return this;
-    }
```
이 부분이 제거된 이유를 혹시 알 수 있을까요?

Dae-Hwa replied:
각각의 부분을 초기화해주는 것 까지 `Board`의 역할이고 이걸 모아서 게임 자체를 초기화하는 것이 `ChessGame`의 역할이라 생각했습니다.
지금 다시 생각해보니 이렇게 하면 `Board`의 상태를 외부에서 바꿔주는 느낌이 들기도 하는데, 혹시 적절하지 않게 보이시나요?

ksundong replied:
그냥 궁금해서 여쭤본거였어요 ㅎㅎ
저라면 그냥두긴 했을 것 같은데, 본지 오래돼서 그 때 생각이 어땠는지 기억이 잘 안나네요.

ksundong commented on src/test/java/kr/codesquad/freddie/chess/board/BoardViewTest.java:
```diff
@@ -0,0 +1,95 @@
+    @BeforeEach
+    void setUp() {
+        board = new Board();
+        boardView = new BoardView(board);
```
의존하는 것을 직접 넣어서 통합테스트로 만들기보다는 mock을 이용해서 단위 테스트를 시도해보세요.
조금 어렵다면, 다음 단계에서 시도해봅시다.

ksundong commented on src/test/java/kr/codesquad/freddie/chess/board/BoardViewTest.java:
```diff
@@ -0,0 +1,95 @@
+    }
+
+    @Test
+    @DisplayName("set으로 넣은 뒤 getRepresentation과 getNumberOf로 확인")
```
boardView에서 테스트 하고자하는 것은 board와는 관계 없이 순수하게 boardView만 테스트하는게 좋습니다.
물론 통합 테스트도 필요하지만, 그럴 때는 테스트 클래스를 분리하는 것을 권장합니다.

ksundong commented on src/test/java/kr/codesquad/freddie/chess/game/ChessGameTest.java:
```diff
@@ -0,0 +1,84 @@
+        int rank = color == Color.BLACK ? 8 : 1;
+        PieceFactory pieceFactory = new PieceFactory(color);
+
+        assertThat(board.findPiece("a" + rank)).isEqualTo(pieceFactory.createRook());
+        assertThat(board.findPiece("b" + rank)).isEqualTo(pieceFactory.createKnight());
+        assertThat(board.findPiece("c" + rank)).isEqualTo(pieceFactory.createBishop());
+        assertThat(board.findPiece("d" + rank)).isEqualTo(pieceFactory.createQueen());
+        assertThat(board.findPiece("e" + rank)).isEqualTo(pieceFactory.createKing());
+        assertThat(board.findPiece("f" + rank)).isEqualTo(pieceFactory.createBishop());
+        assertThat(board.findPiece("g" + rank)).isEqualTo(pieceFactory.createKnight());
+        assertThat(board.findPiece("h" + rank)).isEqualTo(pieceFactory.createRook());
```
되도록이면 assertAll로 묶어주는 편이 개발시에 더 편한 것 같아요.

ksundong commented on src/test/java/kr/codesquad/freddie/chess/game/ChessGameTest.java:
```diff
@@ -0,0 +1,84 @@
+    void initialize() {
+        chessGame.initialize();
+        Board board = chessGame.getBoard();
+        checkInitializeRoyal(Color.BLACK, board);
+        checkInitializeRoyal(Color.WHITE, board);
+        checkInitializePawn(Color.BLACK, board);
+        checkInitializePawn(Color.WHITE, board);
```
여기도 assertAll로 묶어봐주세요~

ksundong commented on src/test/java/kr/codesquad/freddie/chess/game/ChessGameTest.java:
```diff
@@ -0,0 +1,84 @@
+
+        assertThatThrownBy(() -> chessGame.move("b3", "b4"))
+                .isInstanceOf(IllegalArgumentException.class)
+                .hasMessage("이동 위치의 기물이 같은 색상입니다. source : Piece{color=BLACK, kind=KING}, target : Piece{color=BLACK, kind=PAWN}");
```
`toString()` 오버라이딩이 여기 쓰였군요. 👍

ksundong replied:
`assertThatIllegalArgumentException()` 요것도 사용해보세요~

ksundong commented on src/test/java/kr/codesquad/freddie/chess/piece/BishopTest.java:
```diff
@@ -0,0 +1,99 @@
+
+    @Test
+    void isCanMove() {
+        checkIsCanMove("d4", "e5");
+        checkIsCanMove("d4", "f6");
+        checkIsCanMove("d4", "e3");
+        checkIsCanMove("d4", "f2");
+        checkIsCanMove("d4", "c5");
+        checkIsCanMove("d4", "b6");
+        checkIsCanMove("d4", "c3");
+        checkIsCanMove("d4", "b2");
```
이렇게 하기보다는 ParameterizedTest를 시도해봐주세요~
https://www.baeldung.com/parameterized-tests-junit-5
참고자료입니다~

Dae-Hwa replied:
그레이들에  의존성이 없는걸로 알고 있는데 따로 추가해서 사용해도 괜찮을까요?

ksundong replied:
넵 근데 여기 적용하기 좀 힘들지 않을까 싶네요 ㅎㅎ
여유가 되신다면 진행헤주세요~

Dae-Hwa replied:
넵 알겠습니다. 다른 리뷰도 마저 읽어보겠습니다. !! 감사합니다

ksundong commented on src/test/java/kr/codesquad/freddie/chess/utils/PositionCalculatorTest.java:
```diff
@@ -0,0 +1,105 @@
+    }
+
+    @Test
+    void getFileDistance() {
+        positionCalculator = new PositionCalculator(Position.of("a1"), Position.of("b2"));
+
+        assertThat(positionCalculator.getFileDistance()).isEqualTo(-1);
+    }
+
+    @Test
+    void getFileDistanceZero() {
+        positionCalculator = new PositionCalculator(Position.of("a1"), Position.of("a2"));
+
+        assertThat(positionCalculator.getFileDistance()).isEqualTo(0);
+    }
+
+    @Test
+    void getRankDistance() {
+        positionCalculator = new PositionCalculator(Position.of("a1"), Position.of("b2"));
+
+        assertThat(positionCalculator.getRankDistance()).isEqualTo(-1);
+    }
+
+    @Test
+    void getRankDistanceZero() {
+        positionCalculator = new PositionCalculator(Position.of("a1"), Position.of("b1"));
+
+        assertThat(positionCalculator.getRankDistance()).isEqualTo(0);
+    }
```
얘도 ParameterizedTest 적용해봐주세요~

ksundong reviewed (changes requested):
안녕하세요! 프레디! 리뷰어 디온입니다!
구현하느라 정말 고생많으셨을 것 같아요. 저도 리뷰를 하는데 꽤 시간이 걸렸는데 아마 코딩을 하는데에는 더 많은 시간이 소요되고, 많은 고민을 하셨을 것 같습니다. 그런 고민이 많이 느껴지는 코드였어요. 👍 

제가 드릴 피드백은 많지는 않고, 생각해볼만한 것들에 대한 코멘트 위주로 작성했습니다.
요구사항은 충분히 만족하셨다고 판단되고, 그냥 피드백을 확인해보시라는 의미에서 Request changes를 남겨드립니다.
피드백은 다음 미션을 진행하셔야 하니, 반영하지 않으셔도 됩니다!

궁금하시거나, 피드백에 대해서 더 얘기해보시고 싶은 부분이 있으시면, 언제든지 코멘트 남겨주시구요. 가급적이면 하루 내에 답변드릴 수 있도록 하겠습니다!

체스미션을 수행하느라 고생 많으셨습니다.🎉 🎉 🎉 🎉

Dae-Hwa committed:
Change local variables in Board that is not appropriate

Dae-Hwa committed:
Remove unnecessary method in PositionCalculator

Dae-Hwa committed:
Remove unnecessary toString

Dae-Hwa committed:
Remove unnecessary imports

Dae-Hwa committed:
Fix wrong tests of ChessGame

Dae-Hwa committed:
Rename isCanMove to canMove in Piece and its subclasses and apply them

Dae-Hwa committed:
Create Files and use at Board

Dae-Hwa committed:
Remove unnecessary methods

현재 사용하지 않는 메소드 제거

Dae-Hwa committed:
Remove magic number in Move

Dae-Hwa committed:
Extract check logics

Dae-Hwa committed:
Refactor for extracting try-catch block in Main

Dae-Hwa committed:
Create ChessGamePrinter that using at Main

Dae-Hwa committed:
Add const EMPTY_PIECE in Piece

Dae-Hwa committed:
Fix the Pawn can't check moves forward correctly

2칸으로 이동은 폰의 시작 위치에서만 가능.
각 시작위치에서만 2칸 이동할 수 있도록 했는데, 반대 방향으로 2칸 이동할 경우 체스 판의 범위를 벗어나므로 애초에 이동 불가능함.

ksundong commented on src/main/java/kr/codesquad/freddie/chess/Main.java:
```diff
@@ -9,16 +11,24 @@
     public static void main(String[] args) throws IOException {
         try (BufferedReader br = new BufferedReader(new InputStreamReader(System.in))) {
-            Board board = new Board();
+            ChessGame chessGame = ChessGame.create();
+            ChessGamePrinter chessGamePrinter = ChessGamePrinter.create(chessGame);
```
`ChessGamePrinter`가 `chessGame`을 인자로 받는 구조는 조금 이상하다는 생각이 드네요.

ksundong replied:
저라면 `Printer`가 상태를 갖지 않아도 되지 않을까? 라는 생각을 해볼 것 같아요.

Dae-Hwa committed:
Add printer method for err occurred in ChessGamePrinter

ksundong commented on src/main/java/kr/codesquad/freddie/chess/board/Board.java:
```diff
@@ -12,47 +12,28 @@
-        set(destination, findPiece(source));
-        set(source, Piece.createBlank());
+    public void initializeEmptyPiece() {
+        files.fillWithBlankAt(new RankIndex(3).getRankIndexForList());
+        files.fillWithBlankAt(new RankIndex(4).getRankIndexForList());
+        files.fillWithBlankAt(new RankIndex(5).getRankIndexForList());
+        files.fillWithBlankAt(new RankIndex(6).getRankIndexForList());
```
3,4,5,6은 하드 코딩인 것 같은데, 좋은 방법 없을까요?

ksundong commented on src/main/java/kr/codesquad/freddie/chess/board/Files.java:
```diff
@@ -0,0 +1,79 @@
+    }
+
+    public void add(Piece piece) {
+        files.stream()
+
+                .filter(File::isAddable)
```
여기는 사회적 거리두기를 하고 있네요 😅

ksundong commented on src/main/java/kr/codesquad/freddie/chess/board/Files.java:
```diff
@@ -0,0 +1,79 @@
+    }
+
+    public void set(Position position, Piece piece) {
+        files.get(position.getRankIndexForList())
+                .set(position.getFileIndexForList(), piece);
```
조금 verbose한 느낌이 없잖아 있는 것 같아요. `ForList` 때문인 것 같습니다.

ksundong commented on src/main/java/kr/codesquad/freddie/chess/board/Files.java:
```diff
@@ -0,0 +1,79 @@
+                .sum();
+    }
+
+    public Map<CalculablePiece, Double> groupingByCalculablePiece(Color color) {
+        return files.stream()
+                .flatMap(file -> file.getCalculablePieces().stream())
+                .filter(calculablePiece -> calculablePiece.getColor() == color)
+                .collect(Collectors.groupingBy(Function.identity(),
+                        Collectors.summingDouble(CalculablePiece::getPoint)
+                ));
+    }
```
👍 
stream을 아주 잘 활용하셨네요.

ksundong commented on src/main/java/kr/codesquad/freddie/chess/command/Move.java:
```diff
@@ -0,0 +1,26 @@
+
+public class Move implements Command {
+    private static final int COMMAND_LENGTH = 3;
+    private String[] command;
```
얘도 객체로 만들어보는 방법도 있을 것 같네요.
[1], [2]로 사용하는 것 보단 필드로 쓰는게 낫겠죠?

ksundong commented on src/main/java/kr/codesquad/freddie/chess/utils/ChessGamePrinter.java:
```diff
@@ -0,0 +1,30 @@
+import kr.codesquad.freddie.chess.board.BoardView;
+import kr.codesquad.freddie.chess.game.ChessGame;
+
+public class ChessGamePrinter {
```
굳이 상태를 가지기보다는 유틸클래스 느낌으로 상태를 가지지 않도록 해보는 방향도 고민해봐주세요~

ksundong reviewed (approved):
웹 구현안하시고 이거 하셔도 괜찮은건가요 ㅋㅋ...
일단 제가 보기엔 충분히 많은 경험을 하셨다고 판단되어
몇 가지 생각해볼거리 더 남겨드리고 Approve하겠습니다~ 체스게임 구현하느라 고생많으셨습니다!

Dae-Hwa commented:
> 웹 구현안하시고 이거 하셔도 괜찮은건가요 ㅋㅋ...
> 일단 제가 보기엔 충분히 많은 경험을 하셨다고 판단되어
> 몇 가지 생각해볼거리 더 남겨드리고 Approve하겠습니다~ 체스게임 구현하느라 고생많으셨습니다!

오늘안에 다 처리해보려고 했는데 생각보다 속도가 안나서 다 처리하지 못했네요 😓
다시 보니 명확하지 않게 질문한 것도 있었던 것 같은데 자세하게 대답해주셔서 감사합니다!
할때마다 많은 생각이 들었던 부분들이 나와서 질문도 많아졌던 것 같은데 두고두고 생각해보겠습니다. 감사합니다!
